### PR TITLE
Add pre-commit hooks

### PR DESCRIPTION
**Pull request by jrbourbeau**
This PR adds support for (optionally) using [pre-commit](https://pre-commit.com/) to setup and manage pre-commit hooks when developing on Numba. Other projects, like [pandas](https://github.com/pandas-dev/pandas/blob/master/.pre-commit-config.yaml) and [dask](https://github.com/dask/dask/blob/master/.pre-commit-config.yaml), have added support for pre-commit hooks which will automatically run things like `flake8` when trying to `git commit`. I personally have found this very useful and wanted to see if others are also interested in using them here.

### CODE REVIEW

File: docs/source/developer/contributing.rst [hunk 1]
```diff
@@ -214,8 +214,25 @@ Coding conventions
 
 All Python code should follow :pep:`8`.  Our C code doesn't have a
 well-defined coding style (would it be nice to follow :pep:`7`?).
-Code and documentation should generally fit within 80 columns, for
-maximum readability with all existing tools (such as code review UIs).
```
[the remark this answers is not in the document]
Comment (sklam) [reply docs/source/developer/contributing.rst]: Right. I noticed `.flake8` set max line length to 120. We should fix that instead. However, there are so many violations.

[the remark this answers is not in the document]
Comment (jrbourbeau) [reply docs/source/developer/contributing.rst]: I agree the line length in `contributing.rst` and `.flake8` should definitely be consistent. Given that the CI has been enforcing 120 characters per line, and all the violations that occur when making the 120 -> 80 change, is there a reason to prefer 80 over 120?

[the remark this answers is not in the document]
Comment (stuartarchibald) [reply docs/source/developer/contributing.rst]: I have an opinion about this item :) Given the history of slow progression towards PEP8 (still < 50% of files comply) I would not consider anything in `.flake8` as immutable. 

As to the 80 vs 120 question, I prefer 80.

Travis CI logs wrap at ~100, e.g. https://travis-ci.org/numba/numba/jobs/583870284 (97 chars seems to be max), Azure also at ~100 e.g. https://dev.azure.com/numba/numba/_build/results?buildId=2503 (96 chars before the window has to scroll). You can also fit 2x (editor of choice) panes side by side more easily at 80 chars (important for those that use vertical monitors due to the necessity of staring at reams of assembly). Shorter lines tends to also follow Numba's Python use, names tends not to be verbose, heavy OO (e.g. method chaining) is not common.

Also, I think @sklam and I have been doing 80 for quite some time, there's only 50ish files in the code base that are under flake8 control and violating 80 chars, which is about 1/7.

Other opinions are of course welcomed :)

Reference: https://www.python.org/dev/peps/pep-0008/#maximum-line-length (CPython goes for 79)

**Comment by esc** [comment]
Someone proposing a patch means there is at least one person who finds this useful, hence I would recommend to merge this. As for line-length, I am an 80(s) person myself.

**Comment by sklam** [comment]
I like 80, too.  Once the https://github.com/numba/numba/pull/4501#discussion_r325279810 line is restored, I am happy to approve this patch.

**Comment by jrbourbeau** [comment]
Sounds good, thanks for the feedback all. I'll fix the existing files with > 80 characters per line and then update `.flake8` to enforce 80 characters.

**Comment by stuartarchibald** [comment]
@jrbourbeau please could you resolve the minor conflict that's preventing merge, then I think this just needs a quick skim over (as it's just changing line breaks) and it can go in?

**Comment by jrbourbeau** [comment]
Just resolved the merge conflict and fixed a few new line length errors. A quick skim would be great!

File: numba/tests/test_ir_inlining.py [hunk 2]
```diff
@@ -732,7 +732,7 @@ def impl(x):
         def impl():
             a = bar(1)  # integer literal, should inline
             b = bar(2.3)  # float literal, should not inline
-            c = bar(3j)  # complex literal, should inline by virtue of cost model
+            c = bar(3j)  # complex literal, should inline via cost model
```
Comment (stuartarchibald) [inline numba/tests/test_ir_inlining.py]: I think this change is incorrect, it is by virtue of the cost model's decision that the inlining takes place, the cost model does not take part in the inlining itself.

File: numba/tests/test_dispatcher.py [hunk 3]
```diff
@@ -397,7 +400,7 @@ def foo(x):
 
         self.assertIs(foo, foo_rebuilt)
 
-        # do we get the same object even if we delete all the explict references?
+        # do we get the same object even if we delete all explict references?
```
Comment (stuartarchibald) [inline numba/tests/test_dispatcher.py]: I think `the` should be left in

File: numba/tests/test_dispatcher.py [hunk 4]
```diff
@@ -1312,7 +1320,7 @@ def test_frozen(self):
             self.assertIsNone(locator)
 
             sys.frozen = True
-            # returns a cache locator object, only works when executable is frozen
+            # returns cache locator object, only works when executable is frozen
```
Comment (stuartarchibald) [inline numba/tests/test_dispatcher.py]: I think the `a` should be left in, and if anything this perhaps ought to read:
```
 # returns a cache locator object, only works when the executable is frozen
```

File: numba/tests/test_np_functions.py [hunk 5]
```diff
@@ -1356,19 +1363,20 @@ def test_triu_indices_from(self):
...
 
     def partition_sanity_check(self, pyfunc, cfunc, a, kth):
-        # as NumPy uses a different algorithm, we do not expect to match outputs exactly...
+        # as NumPy uses a different algorithm, we do not expect to
+        # match outputs exactly...
         expected = pyfunc(a, kth)
         got = cfunc(a, kth)
 
-        # ... but we do expect the unordered collection of elements up to kth to tie out
+        # but we do expect unordered collection of elements up to kth to tie out
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]:
```suggestion
        # but we do expect the unordered collection of elements up to the kth to tie out
```
this needs more definite article (and probably a line wrap).

File: numba/tests/test_np_functions.py [hunk 6]
```diff
@@ -1356,19 +1363,20 @@ def test_triu_indices_from(self):
...
         self.assertPreciseEqual(np.unique(expected[:kth]), np.unique(got[:kth]))
 
-        # ... likewise the unordered collection of elements from kth onwards
+        # likewise the unordered collection of elements from kth onwards
```
Comment (stuartarchibald) [inline numba/tests/test_np_functions.py]:
```suggestion
        # likewise the unordered collection of elements from the kth onwards
```

File: numba/tests/test_unicode.py [hunk 7]
```diff
@@ -741,7 +747,8 @@ def test_join_non_string_exception(self):
         # Handle empty separator exception
         with self.assertRaises(TypingError) as raises:
             cfunc('', [1, 2, 3])
-        # This error message is obscure, but indicates the error was trapped in typing of str.join()
+        # This error message is obscure, but indicates the error was trapped
+        # in typing of str.join()
```
Comment (stuartarchibald) [inline numba/tests/test_unicode.py]:
```suggestion
        # in the typing of str.join()
```

**Pull request merged**